_posts/2018-06-25-refactoring-with-design-patterns.markdown

```diff
@@ -0,0 +1,194 @@
+
+  def self.calculate_net_salary(country, amount)
+    country_taxes = case country
+      when "Argentina"
+        (amount * ARG_TAX)
+        # Argentina tax calculation
+      when "USA"
+        (amount * USA_TAX) + 200
+        # USA tax calculation
+      when "Brazil"
+        (amount + 500) / BRA_TAX
+        # Brazil tax calculation
+
+    amount - taxes
+  end
+end
+```
+
+We can easily see that the `calculate_net_salary` will massively grow each time that we need to add a new country. This situation breaks the [open/closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) that says: *"software entities (classes, modules, functions, etc.) should be open for extension, but closed for modification"*. This means that to make a class do new things you shouldn’t need to change the class itself.
+
+If we can not extend the `calculate_net_salary` method to make a new country calculation without a lot of modification we are not following this principle. To solve this situation we can apply the Strategy Pattern to refactor our code.
+
+The <strong>Strategy</strong> is a behavioral design pattern that suggests to take a class that does something important in a lot of different ways and extract all these algorithms into separate classes called strategies. The original class, called *context*, will receive a field that references to one of the strategies.
+
+With that concept in mind we can refactor the context class:
+
+```ruby
+# lib/strategy_sample/employee_salary.rb
+class EmployeeSalary
+
+  def initialize(strategy)
+    @strategy = strategy
+  end
+
+  def initialize
+    @strategies = {
+      'USA': UsaTaxes,
+      'ARG': ArgentinaTaxes,
+      'BRA': BrazilTaxes
+    }
+  end
+
+  def self.calculate_net_salary(amount, country)
+    strategy = @strategies[country]
+
+    strategy ? amount - strategy.taxes(amount) : amount
+  end
+end
+```
+
+Then we will create one strategy for each one of the necessary country calculations.
+
+```ruby
+# lib/strategy_sample/strategies/usa_taxes.rb
+class UsaTaxes
+  def self.taxes(amount)
+    # USA tax calculation here
+  end
+end
+```
+
+```ruby
+# lib/strategy_sample/strategies/argentina_taxes.rb
+class ArgentinaTaxes
+  def self.taxes(amount)
+    # Argentina tax calculation here
+  end
+end
+```
+
+```ruby
+# lib/strategy_sample/strategies/argentina_taxes.rb
+class BrazilTaxes
+  def self.taxes(amount)
+    # Argentina tax calculation here
+  end
+end
+```
+
+That way the single responsibility of the `EmployeeSalary` class is delegate the calculation work to a linked strategy instead of executing it on its own.
+
+Wherever we need to get the net salary we should call the `EmployeeSalary` class like this:
+
+```ruby
+Taxes.new.net_salary(1000, "BRA")
+```
+
+Now if we need to add a new country we just need to create a new strategy and add to the context class. With this refactor we can keep all the concerns separated and the code is cleaner and easier to maintain.
```

Comment:
@cleicar Great article! Maybe you can add a brief summary/conclusion at the end of the post?